File: 02 - INTRODUÇÃO AOS ALGORITMOS E PROGRAMAÇÃO/praticas/prova_respondida_por_mim/Questao9.php
<?php

// Questão 9

// Algumas pessoas estão em uma fila no parque.
// As pessoas e as arvores no parque serão representadas
// por um vetor.

// O valor -1 representa uma árvore
// valor positivo representa a altura de uma pessoa

// Existem árvores entre as pessoas e as árvores 
// não não podem ser movidas.

// Sua tarefa é reorganizar as pessoas por suas alturas
// em uma ordem não decrescente, sem mover as árvores.

// Considere que as pessoas podem ser muito altas!

//Exemplo

//Se $a = [-1, 150, 190, 170, -1, -1, 160, 180], 
// o retorno deve ser [-1, 150, 160, 170, -1, -1, 180, 190].

// $a = [];

function sortByHeight($a)
{
    $a = [-1, 150, 190, 170, -1, -1, 160, 180];
    $personsArr = [];

    foreach ($a as $item){
        if ($item != -1)
            $personsArr[] = $item;
    }

    sort($personsArr);
    $i =0;
    foreach ($a as $key=>$item){
        if ($item != -1)
            $a[$key] = $personsArr[$i++];
    }

    return $a;
}

echo "RESPOSTA [1, 3, -1, 23, 43, -1, -1, 54, -1, -1, -1, 77]";

// RESPOSTA = [1, 3, -1, 23, 43, -1, -1, 54, -1, -1, -1, 77]

File: 02 - INTRODUÇÃO AOS ALGORITMOS E PROGRAMAÇÃO/praticas/prova_respondida_por_mim/Questao 9.py


def sortByHeight(a):
    b = sorted([i for i in a if i!=-1])
    j = 0
    for i in range(len(a)):
        if a[i] == -1:
            pass
        else:
            a[i]=b[j]
            j+=1
    return a

if __name__ == '__main__':

    a= [23,54,-1,43,1,-1,-1,77,-1,-1,-1,3]

    print (sortByHeight(a))

    # RESPOSTA = [1, 3, -1, 23, 43, -1, -1, 54, -1, -1, -1, 77] ✅